### DISCUSSION TITLE
OLAP, OLTP에 대한 개념 및 차이

### DISCUSSION THREAD

**Discussion by GIVEN53** [2024-woowacourse-study/level-interview#192] [Database]
.

**Comment by Kimprodp** [comment]
**[OLTP (Online Transaction Processing)]**
- 실시간으로 DB에 트랜잭션을 처리하는 시스템
- 실시간 데이터 처리와 무결성을 중시하며, 주로 일상적인 운영 업무에서 사용
- 복수의 사용자 PC에서 발생되는 트랜잭션(Transaction)을 DB 서버가 처리, 그 결과를 요청한 사용자에게 결과값을 되돌려주는 과정
- 1개의 트랜잭션에서 발생되는 INSERT, UPDATE, DELETE를 무결성 보장하여 처리, 결과를 SELECT하는 과정

**[OLAP (Online Analytical Processing)]**
- 대량의 데이터를 분석하고 복잡한 쿼리를 처리하는 시스템
- 주로 비즈니스 인텔리전스와 데이터 분석에서 사용
- 데이터 웨어하우스 (DW), DB에 저장되어 있는 데이터를 분석, 사용자에게 유의미한 정보를 제공해주는 처리 방법
- 기존에 저장되어 있는 데이터를 사용자의 요구와 목적에 맞게 분석하여 정보를 제공하는 개념

**[OLAP와 OLTP 차이점]**
- OLAP 데이터베이스 아키텍처는 데이터 쓰기 작업보다 데이터 읽기 작업에 우선 순위를 두어 대용량 데이터를 기반으로 복잡한 쿼리를 빠르고 효율적으로 수행할 수 있다. 반면에 OLTP 데이터베이스 아키텍처는 데이터 쓰기 작업에 우선 순위를 두어 데이터 무결성을 손상시키지 않으면서 대용량 트랜잭션 데이터를 잦은 빈도로 업데이트 할 수 있다.
- OLAP 데이터베이스를 업데이트할 때는 정기적으로 데이터를 대규모 배치로 처리한 다음 이 배치를 한 번에 시스템에 업로드 하여 비교적 긴 응답 시간을 가진다. 반대로 OLTP는 데이터베이스 업데이트를 실시간으로 관리하여 업데이트가 빠르다.

**Comment by seunghye218** [comment]
## OLTP

OLTP(Online Transactional Processing)는 ATM 및 온라인 뱅킹, 금전 등록기, 전자상거래 그리고 우리가 매일 상호작용하는 수많은 기타 서비스를 위한 신속하고 정확한 데이터 처리를 가능하게 합니다.

OLTP(Online Transactional Processing)는 일반적으로 인터넷을 통해 많은 사람들이 많은 데이터베이스 트랜잭션을 실시간으로 실행할 수 있도록 지원합니다.

데이터베이스 트랙잭션은 데이터베이스에서 데이터를 변경, 삽입, 삭제 또는 쿼리하는 것을 말합니다. OLTP 시스템(그리고 OLTP 시스템이 지원하는 데이터베이스 트랜잭션)은 온라인 뱅킹 및 ATM 트랜잭션, 전자상거래 및 오프라인 구매, 호텔 및 항공권 예약 등 우리가 매일 수행하는 금융 거래 중 다수를 지원합니다. 각각의 경우에 데이터베이스 트랜잭션은 또한 해당 금융 트랜잭션의 레코드로 남아 있습니다. OLTP는 또한 비밀번호 변경 및 문자 메시지와 같은 비금융 데이터베이스 교환도 지원합니다.

OLTP에서 모든 데이이터베이스 트랜잭션을 정의하는 공통된 특징은 *원자성* (또는 분할 불가능성)입니다. 즉, 트랜잭션은 전체로서 성공하거나 실패합니다(또는 취소됩니다). 트랜잭션은 보류 또는 중간 상태에 머물 수 없습니다.

## OLAP

OLAP 또는 온라인 분석 처리는 데이터 웨어하우스, 데이터 레이크 또는 기타 데이터 저장소에 있는 대량의 데이터의 복잡한 쿼리 또는 다차원 분석을 고속으로 수행하기 위한 기술입니다. OLAP는 비즈니스 인텔리전스(BI), 의사 결정 지원, 다양한 비즈니스 예측 및 보고 응용 프로그램 등에 이용됩니다.

대부분의 비즈니스 데이터에는 여러 차원, 즉 프레젠테이션, 추적 또는 분석을 위해 데이터가 분류되는 여러 범주가 있습니다. 예를 들어 판매 수치에는 위치(지역, 국가, 시/도, 매장), 시간(일, 주, 월, 년), 제품(의류, 남성/여성/아동, 브랜드, 유형) 등과 관련된 여러 차원이 있을 수 있습니다.

그러나 데이터 세트는 데이터 웨어하우스나 데이터 레이크에서 테이블에 저장되므로 각 테이블은 데이터를 한 번에 두 개의 차원으로만 구성할 수 있습니다. OLAP는 여러 관계형 데이터 세트에서 데이터를 추출하여 이를 신속한 처리와 통찰력 있는 분석이 가능한 다차원 형식으로 재구성합니다.

## OLAP 및 OLTP 비교

OLAP와 OLTP의 주요 차이점은 이름에 있습니다. OLAP는 본질적으로 분석적이고 OLTP는 트랜잭션입니다.

OLAP 도구는 트랜잭션 데이터와 기록 데이터를 모두 포함하는 데이터 웨어하우스의 데이터를 다차원적으로 분석하도록 설계되었습니다. 실제로 OLAP 서버는 일반적으로 데이터 웨어하우징 솔루션의 중간 분석 계층입니다. OLAP의 일반적인 용도로는 데이터 마이닝 및 기타 비즈니스 인텔리전스 애플리케이션, 복잡한 분석 계산 및 예측 시나리오뿐만 아니라 재무 분석, 예산 책정 및 예측 계획과 같은 비즈니스 보고 기능이 있습니다.

OLTP는 최근 트랜잭션을 가능한 한 빠르고 정확하게 처리하여 트랜잭션 지향 애플리케이션을 지원하도록 설계되었습니다. OLTP의 일반적인 용도로는 ATM, 전자 상거래 소프트웨어, 신용 카드 결제 처리, 온라인 예약, 예약 시스템 및 기록 보관 도구가 있습니다.

**Comment by 3Juhwan** [comment]
OLTP 시스템 | OLAP 시스템
-- | --
다수의 사용자에 의한 대량의 데이터베이스 트랜잭션을 실시간으로 실행할 수 있도록 지원 | 일반적으로 분석을 목적으로 데이터베이스 내 다수의 레코드(또는 모든 레코드)에 대한 질의 작업 포함
빛의 속도에 가까운 빠른 응답시간 필요 | OLTP에 대비 엄청나게 느린 응답시간 필요
적은 양의 데이터를 자주 수정하고, 일반적으로 읽기 및 쓰기 작업 간 균형이 유지됨 | 데이터를 전혀 수정하지 않음. 일반적으로 읽기 집약적인 작업
인덱스화된 데이터를 사용해 응답 시간 개선 | 대량의 레코드에 손쉽게 액세스할 수 있도록 컬럼 형식으로 데이터 저장
데이터베이스에 대한 빈번한 또는 동시 백업 필요 | 훨씬 적은 빈도의 데이터베이스 백업 필요
상대적으로 적은 스토리지 공간 필요 | 대량의 기록 데이터를 저장하기 때문에 일반적으로 상당한 양의 스토리지 공간 필요
일반적으로 하나 또는 몇 개의 레코드를 포함하는 단순한 쿼리 실행 | 다수의 레코드를 포함하는 복잡한 쿼리 실행

출처: https://www.oracle.com/kr/database/what-is-oltp/
